frontend/src/components/common/PartySelectField.js

```diff
@@ -0,0 +1,187 @@
+    </a>
+  </div>
+);
+
+const transformData = (data, options, footer, searchTerm) => {
+  const transformedData = data
+    .map((opt) => (
+      <AutoComplete.Option key={opt} value={opt}>
+        {options[opt].name}
+      </AutoComplete.Option>
+    ))
+    .sort(
+      (a, b) =>
+        compareTwoStrings(searchTerm, b.props.children) -
+        compareTwoStrings(searchTerm, a.props.children)
+    );
+  return footer
+    ? transformedData.concat(
+        <AutoComplete.Option disabled key="footer" value="footer">
+          {footer}
+        </AutoComplete.Option>
+      )
+    : transformedData;
+};
+
+export class PartySelectField extends Component {
+  state = { selectedOption: { key: "", label: "" }, partyDataSource: [] };
+
+  constructor(props) {
+    super(props);
+    this.fetchPartiesDebounced = debounce(this.props.fetchParties, 1000);
+  }
+
+  showAddPartyForm = () => {
+    this.props.setAddPartyFormState({
+      showingAddPartyForm: true,
+      person: !this.props.person,
+      organization: !this.props.organization,
+      partyLabel: this.props.partyLabel,
+    });
+  };
+
+  componentWillReceiveProps = (nextProps) => {
+    const partiesUpdated = this.props.parties !== nextProps.parties;
+    const lastCreatedPartyUpdated = this.props.lastCreatedParty !== nextProps.lastCreatedParty;
+
+    if (partiesUpdated) {
+      let filteredParties = nextProps.parties;
+
+      if (this.props.organization && !this.props.person) {
+        filteredParties = filteredParties.filter(
+          ({ party_type_code }) => party_type_code === "ORG"
+        );
+      } else if (this.props.person && !this.props.organization) {
+        filteredParties = filteredParties.filter(
+          ({ party_type_code }) => party_type_code === "PER"
+        );
+      }
+
+      this.setState((prevState) => {
+        const newPartyDataSource = transformData(
+          createItemIdsArray(filteredParties, "party_guid"),
+          createItemMap(filteredParties, "party_guid"),
+          this.props.allowAddingParties &&
+            renderAddPartyFooter(this.showAddPartyForm, this.props.partyLabel),
+          prevState.selectedOption.label
+        );
+        return { partyDataSource: newPartyDataSource };
+      });
+
+      if (lastCreatedPartyUpdated) {
+        this.setState({
+          selectedOption: {
+            key: nextProps.lastCreatedParty.party_guid,
+            label: nextProps.lastCreatedParty.name,
+          },
+        });
+      }
+    }
+  };
+
+  handleSearch = (value) => {
+    this.fetchPartiesDebounced({ name_search: value });
+    this.setState({ selectedOption: { key: value, label: value } });
+  };
+
+  handleSelect = (value, option) => {
+    this.setState({ selectedOption: option });
+  };
+
+  validOption = (value) =>
+    this.state.partyDataSource.find((opt) => opt.key === value)
+      ? undefined
+      : `"Invalid ${this.props.partyLabel}`;
+
+  render = () => (
+    <Field
+      {...this.props}
```

Comment:
It's to pass any desired <Field/> prop through on the form level, of which there are a lot and I thought including them would just clutter this class. The intention is to have <PartySelectField/> behave as any one would expect any other <Field/> would. 

I really wanted this to be used as a component={renderConfig.PartySelect} but the complexity of AutoComplete's own internal behavior made using a controlled value not seem to work with <Field/>'s event hooks. I'll revisit that at some point because it should work, and then this wouldn't be needed.

Reply:
Perfect. I figured that the reason was going to be something like that. I'll mark this as resolved